src/Analyser/VariableTypeHolder.php

```diff
@@ -34,10 +36,17 @@ public function equals(self $other): bool
 
 	public function and(self $other): self
 	{
-		if ($this->getType()->equals($other->getType())) {
-			$type = $this->getType();
+		$thisType = $this->getType();
+		$otherType = $other->getType();
+
+		if ($thisType->equals($otherType)) {
+			$type = $thisType;
```

Comment:
I have the feeling that union should deal with this in a more general approach. But without loosing the other types completely :)